### DISCUSSION TITLE
How expensive are topics?

### DISCUSSION THREAD

**Discussion by gmencz** [uNetworking/uWebSockets.js#558] [Q&A] [answered]
Are topics expensive or not? Could I use millions of topics (topic per user for example) without trouble?

**Comment by ghost** [comment, marked as answer]
Topics are expensive and not suited for per-user

**Reply by gmencz** [reply]
I see, do you have any ideas for how to implement user "presence"? I was thinking about having a topic per-user and that way I can check for the number of subscribers in that topic and if it's >0 then that means the user is online but can't do this since topics are expensive sadly.

**Reply by ghost** [reply]
A Set? That's literally a Set. Use a Set.

**Reply by gmencz** [reply]
Okay, I always overthink these kinds of things because "What if I have a million elements in the set and I need to iterate over them, that will be really slow".

**Reply by ghost** [reply]
;D

**Reply by gmencz** [reply]
So, one last question regarding topics being "expensive", would topics be suitable for something like Discord? Where each server is a topic users subscribe to? I'm assuming you're familiar with Discord but if not I'd be more than happy to explain how it would work by using pubsub. 

If using the built in pubsub isn't suitable then I guess I'd keep the channels separate in some Map along with the sockets but then if I have to broadcast to all sockets I'd have to use a for-loop and run into some perf issues when the sockets grow.

**Reply by ghost** [reply]
Yes Discord Server is a much better granularity. 1 per connection is too fine, 1 per Discord Server is fine. but try and load balance connections so that one instance holds only a limited set of topics anyways.

**Reply by gmencz** [reply]
Alright, are there any benchmarks regarding topics? I'm going to benchmark them myself on my machine but it would be nice to have some "general" benchmarks data.
